yarn-project/sequencer-client/src/sequencer/timetable.ts

```diff
@@ -0,0 +1,127 @@
+import { createLogger } from '@aztec/aztec.js';
+
+import { type SequencerMetrics } from './metrics.js';
+import { SequencerState } from './utils.js';
+
+export class SequencerTimetable {
+  /** How late into the slot can we be to start working */
+  public readonly initialTime = 2;
+
+  /** How long it takes to get ready to start building */
+  public readonly blockPrepareTime = 2;
+
+  /** How long it takes to for proposals and attestations to travel across the p2p layer (one-way) */
+  public readonly attestationPropagationTime = 2;
+
+  /** How much time we spend validating and processing a block after building it, and assembling the proposal to send to attestors */
+  public readonly blockValidationTime = 1;
+
+  /**
+   * How long it takes to get a published block into L1. L1 builders typically accept txs up to 4 seconds into their slot,
+   * but we'll timeout sooner to give it more time to propagate (remember we also have blobs!). Still, when working in anvil,
+   * we can just post in the very last second of the L1 slot and still expect the tx to be accepted.
+   */
+  public readonly l1PublishingTime;
+
+  constructor(
+    private readonly ethereumSlotDuration: number,
+    private readonly aztecSlotDuration: number,
+    private readonly maxL1TxInclusionTimeIntoSlot: number,
+    private readonly enforce: boolean = true,
+    private readonly metrics?: SequencerMetrics,
+    private readonly log = createLogger('sequencer:timetable'),
+  ) {
+    this.l1PublishingTime = this.ethereumSlotDuration - this.maxL1TxInclusionTimeIntoSlot;
+  }
+
+  private get afterBlockBuildingTimeNeededWithoutReexec() {
+    return this.blockValidationTime + this.attestationPropagationTime * 2 + this.l1PublishingTime;
+  }
+
+  public getBlockProposalExecTimeEnd(secondsIntoSlot: number): number {
+    // We are N seconds into the slot. We need to account for `afterBlockBuildingTimeNeededWithoutReexec` seconds,
+    // send then split the remaining time between the re-execution and the block building.
+    const maxAllowed = this.aztecSlotDuration - this.afterBlockBuildingTimeNeededWithoutReexec;
+    const available = maxAllowed - secondsIntoSlot;
+    const executionTimeEnd = secondsIntoSlot + available / 2;
+    this.log.debug(`Block proposal execution time deadline is ${executionTimeEnd}`, {
+      secondsIntoSlot,
+      maxAllowed,
+      available,
+      executionTimeEnd,
+    });
+    return executionTimeEnd;
+  }
+
+  private get afterBlockReexecTimeNeeded() {
+    return this.attestationPropagationTime + this.l1PublishingTime;
+  }
+
+  public getValidatorReexecTimeEnd(secondsIntoSlot: number): number {
+    // We are N seconds into the slot. We need to account for `afterBlockReexecTimeNeeded` seconds.
+    const maxAllowed = this.aztecSlotDuration - this.afterBlockReexecTimeNeeded;
+    const available = maxAllowed - secondsIntoSlot;
+    const validationTimeEnd = secondsIntoSlot + available / 2;
```

Comment:
Why is this split? Is it not simply `maxAllowed`?

Reply:
Because I got it wrong. Good catch.

Reply:
Fixed in [0f46b48](https://github.com/AztecProtocol/aztec-packages/pull/11221/commits/0f46b488cea250adb2496256e5be4c34958a7dbc)